File: app/src/main/java/igrek/todotree/intent/NavigationCommand.java
package igrek.todotree.intent;


import org.joda.time.DateTime;

import javax.inject.Inject;

import igrek.todotree.R;
import igrek.todotree.activity.ActivityController;
import igrek.todotree.app.AppData;
import igrek.todotree.app.AppState;
import igrek.todotree.dagger.DaggerIoc;
import igrek.todotree.service.summary.AlarmService;
import igrek.todotree.service.summary.NotificationService;
import igrek.todotree.service.tree.TreeManager;
import igrek.todotree.service.tree.TreeSelectionManager;
import igrek.todotree.ui.GUI;

public class NavigationCommand {
	
	@Inject
	TreeManager treeManager;
	
	@Inject
	GUI gui;
	
	@Inject
	ActivityController activityController;
	
	@Inject
	AppData appData;
	
	@Inject
	TreeSelectionManager selectionManager;
	
	@Inject
	NotificationService notificationService;
	
	@Inject
	AlarmService alarmService;
	
	public NavigationCommand() {
		DaggerIoc.factoryComponent.inject(this);
	}
	
	public boolean optionsSelect(int id) {
		switch (id) {
			case R.id.action_minimize:
				activityController.minimize();
				return true;
			case R.id.action_exit_without_saving:
				new ExitCommand().exitApp();
				return true;
			case R.id.action_save_exit:
				new ExitCommand().optionSaveAndExit();
				return true;
			case R.id.action_save:
				new PersistenceCommand().optionSave();
				return true;
			case R.id.action_reload:
				new PersistenceCommand().optionReload();
				return true;
			case R.id.action_restore_backup:
				new PersistenceCommand().optionRestoreBackup();
				return true;
			case R.id.action_select_all:
				new ItemSelectionCommand().toggleSelectAll();
				return false;
			case R.id.action_cut:
				new ClipboardCommand().cutSelectedItems();
				return false;
			case R.id.action_copy:
				new ClipboardCommand().copySelectedItems();
				return false;
			case R.id.action_sum_selected:
				new ItemSelectionCommand().sumItems();
				return false;
			case R.id.action_show_statistics:
				new StatisticsCommand().showStatisticsInfo();
				return false;
			case R.id.action_go_up:
				new TreeCommand().goUp();
				return false;
			case R.id.action_notify:
				summaryNotify();
				return false;
		}
		return false;
	}
	
	private void summaryNotify() {
		alarmService.setAlarmAt(DateTime.now().plusSeconds(10));
	}
	
	public boolean backClicked() {
		if (appData.isState(AppState.ITEMS_LIST)) {
			if (selectionManager.isAnythingSelected()) {
				selectionManager.cancelSelectionMode();
				new GUICommand().updateItemsList();
			} else {
				new TreeCommand().goBack();
			}
		} else if (appData.isState(AppState.EDIT_ITEM_CONTENT)) {
			if (gui.editItemBackClicked())
				return true;
			new ItemEditorCommand().cancelEditedItem();
		}
		return true;
	}
	
	public boolean approveClicked() {
		if (appData.isState(AppState.EDIT_ITEM_CONTENT)) {
			gui.requestSaveEditedItem();
		} else if (appData.isState(AppState.ITEMS_LIST)) {
			if (selectionManager.isAnythingSelected()) {
				selectionManager.cancelSelectionMode();
				new GUICommand().updateItemsList();
			} else {
				new TreeCommand().goBack();
			}
		}
		return true;
	}
	
}

File: app/src/main/java/igrek/todotree/intent/NavigationCommand.kt
package igrek.todotree.intent

import android.content.Context
import android.os.Handler
import android.os.Looper
import igrek.todotree.R
import igrek.todotree.activity.ActivityController
import igrek.todotree.app.AppData
import igrek.todotree.app.AppState
import igrek.todotree.inject.LazyExtractor
import igrek.todotree.inject.LazyInject
import igrek.todotree.inject.appFactory
import igrek.todotree.command.Commander
import igrek.todotree.layout.navigation.NavigationMenuController
import igrek.todotree.service.import.DatabaseImportFileChooser
import igrek.todotree.service.tree.TreeSelectionManager
import igrek.todotree.ui.GUI
import igrek.todotree.ui.treelist.TreeListLayout

class NavigationCommand(
    context: LazyInject<Context> = appFactory.context,
    gui: LazyInject<GUI> = appFactory.gui,
    appData: LazyInject<AppData> = appFactory.appData,
    activityController: LazyInject<ActivityController> = appFactory.activityController,
    treeSelectionManager: LazyInject<TreeSelectionManager> = appFactory.treeSelectionManager,
    navigationMenuController: LazyInject<NavigationMenuController> = appFactory.navigationMenuController,
) {
    private val context by LazyExtractor(context)
    private val gui by LazyExtractor(gui)
    private val appData by LazyExtractor(appData)
    private val activityController by LazyExtractor(activityController)
    private val treeSelectionManager by LazyExtractor(treeSelectionManager)
    private val navigationMenuController by LazyExtractor(navigationMenuController)
    private val treeListLayout: TreeListLayout by LazyExtractor(appFactory.treeListLayout)

    fun optionsSelect(id: Int): Boolean {
        when (id) {
            R.id.action_minimize -> {
                activityController.minimize()
                return true
            }
            R.id.action_exit_without_saving -> {
                ExitCommand().exitDiscardingChanges()
                return true
            }
            R.id.action_save_exit -> {
                ExitCommand().saveItemAndExit()
                return true
            }
            R.id.action_save -> {
                PersistenceCommand().saveDatabaseUi()
                return true
            }
            R.id.action_reload -> {
                PersistenceCommand().optionReload()
                return true
            }
            R.id.action_restore_backup -> {
                PersistenceCommand().optionRestoreBackup()
                return true
            }
            R.id.action_import_db -> {
                DatabaseImportFileChooser().showFileChooser()
                return true
            }
            R.id.action_select_all -> {
                ItemSelectionCommand().toggleSelectAll()
                return false
            }
            R.id.action_cut -> {
                ClipboardCommand().cutSelectedItems()
                return false
            }
            R.id.action_copy -> {
                ClipboardCommand().copySelectedItems()
                return false
            }
            R.id.action_go_up -> {
                TreeCommand().goStepUp()
                return false
            }
            R.id.action_enter_command -> {
                Commander(context).showCommandAlert()
                return false
            }
            R.id.action_open_drawer -> {
                navigationMenuController.navDrawerShow()
                return false
            }
        }
        return false
    }

    fun backClicked(): Boolean {
        gui.startLoading()
        Handler(Looper.getMainLooper()).post {
            if (appData.isState(AppState.ITEMS_LIST)) {
                if (treeSelectionManager.isAnythingSelected) {
                    treeSelectionManager.cancelSelectionMode()
                    treeListLayout.updateItemsList()
                } else {
                    TreeCommand().goBack()
                }
            } else if (appData.isState(AppState.EDIT_ITEM_CONTENT)) {
                gui.onEditBackClicked()
            }
        }
        return true
    }

    fun doneClicked(): Boolean {
        if (appData.isState(AppState.EDIT_ITEM_CONTENT)) {
            gui.requestSaveEditedItem()
        } else if (appData.isState(AppState.ITEMS_LIST)) {
            if (treeSelectionManager.isAnythingSelected) {
                treeSelectionManager.cancelSelectionMode()
                treeListLayout.updateItemsList()
            } else {
                TreeCommand().goBack()
            }
        }
        return true
    }

    fun doneAllClicked(): Boolean {
        ExitCommand().saveItemAndExit()
        return true
    }
}